homeassistant/components/websocket_api/commands.py

```diff
@@ -231,19 +235,68 @@ async def handle_call_service(
     except ServiceNotFound as err:
         if err.domain == msg["domain"] and err.service == msg["service"]:
-            connection.send_error(msg["id"], const.ERR_NOT_FOUND, "Service not found.")
+            err.translation_key = "service_not_found_general"
+            err.domain = const.DOMAIN
+            code = const.ERR_NOT_FOUND
         else:
-            connection.send_error(msg["id"], const.ERR_HOME_ASSISTANT_ERROR, str(err))
+            err.translation_key = "service_not_found"
+            err.domain = const.DOMAIN
+            code = const.ERR_HOME_ASSISTANT_ERROR
+        message = await async_build_error_message(hass, err)
+        connection.send_error(msg["id"], code, message)
     except vol.Invalid as err:
         connection.send_error(msg["id"], const.ERR_INVALID_FORMAT, str(err))
+    except ServiceValidationError as err:
+        connection.logger.error(err)
+        connection.logger.debug("", exc_info=err)
+        message = await async_build_error_message(hass, err)
+        connection.send_error(msg["id"], const.ERR_HOME_ASSISTANT_ERROR, message)
     except HomeAssistantError as err:
         connection.logger.exception(err)
-        connection.send_error(msg["id"], const.ERR_HOME_ASSISTANT_ERROR, str(err))
+        message = await async_build_error_message(hass, err)
+        connection.send_error(msg["id"], const.ERR_HOME_ASSISTANT_ERROR, message)
     except Exception as err:  # pylint: disable=broad-except
         connection.logger.exception(err)
         connection.send_error(msg["id"], const.ERR_UNKNOWN_ERROR, str(err))
 
 
+async def async_get_exception_translations(
+    hass: HomeAssistant, domain: str
+) -> dict[str, Any]:
+    """Get translations for exceptions for a domain."""
+    return await async_get_translations(
+        hass, hass.config.language, "exceptions", {domain}
+    )
+
+
+async def async_build_error_message(
+    hass: HomeAssistant, err: HomeAssistantError
+) -> str:
+    """Build translated error message from exception."""
+    if (translation_key := err.translation_key) is None or (
+        domain := err.domain
+    ) is None:
+        return str(err)
+
+    exception_translations: dict[str, dict[str, Any]] = hass.data.setdefault(
+        DATA_EXCEPTION_TRANSLATIONS, {}
+    )
+    if (domain_exception_translations := exception_translations.get(domain)) is None:
+        domain_exception_translations = await async_get_exception_translations(
+            hass, domain
+        )
+        exception_translations[domain] = domain_exception_translations
+    exception_message: str | None
+    if (
+        exception_message := domain_exception_translations.get(
+            f"component.{domain}.exceptions.{translation_key}.message",
+        )
+    ) is None:
+        return str(err)
+    place_holders = err.translation_placeholders or {}
+    return exception_message.format(**place_holders, message=str(err))  # type: ignore[no-any-return]
+
+
```

Comment:
As discussed on Discord, this should be implemented in frontend. Instead of translating in core, core should send translation key etc. in the error messages.

Reply:
Removed, this will be planned to be processed in the frontend.